ion-schema/src/main/kotlin/com/amazon/ionschema/model/AstConstraint.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazon.ionschema.model
+
+/**
+ * Represents an Ion Schema Language constraint. Implementations must add any necessary properties for modelling the
+ * details of that particular constraint.
+ *
+ * The recursive type bound enforces that [id] has a valueType that is the same type as the implementation of this interface.
+ *
+ * Sometimes we may need to get a ConstraintId in a static way (i.e. we have no instance of this constraint). For this
+ * reason, it is recommended that all concrete implementations of [AstConstraint] include a companion object like this
+ * example from [AllOfConstraint]:
+ * ```
+ * companion object : ConstraintId<AllOfConstraint> by ConstraintId("all_of") {
+ *     @JvmField val ID = this@Companion
+ * }
+ * ```
+ * The companion object implements [ConstraintId] which means that when using this library from Kotlin, the class
+ * itself is like a static [ConstraintId].
+ * ```
+ * val someId: ConstraintId<*> = AllOfConstraint
+ * ```
+ * When using this library from Java, the ConstraintId can be accessed from `AllOfConstraint.ID`
+ * ```
+ * ConstraintId<?> someId = AllOfConstraint.ID;
+ * ```
+ */
+interface AstConstraint<T : AstConstraint<T>> {
```

Comment:
I find the use of `Ast` prefixing for this and others a bit confusing. It's really just the schema as domain model.

I'm also slightly skeptical of the value of having the interface 'have-an' `id` for itself. I see how that enforces what you want, but it comes at the cost of the cognitive load of the self referential type parameter. I think you could just establish a convention instead, I may be wrong.

Reply:
> I find the use of Ast prefixing for this and others a bit confusing. It's really just the schema as domain model.

It is a domain model, but that isn't mutually exclusive with it being an abstract syntax tree when you're writing a programming language interpreter. (That's essentially what an Ion Schema System is.)

However, I agree that the prefix is unnecessary. 

> I'm also slightly skeptical of the value of having the interface 'have-an' id for itself. I see how that enforces what you want, but it comes at the cost of the cognitive load of the self referential type parameter. I think you could just establish a convention instead, I may be wrong.

Are you referring to the static or instance variable?
* The instance variable is necessary because the constraint implementation class is not enough to identify a constraint (that is assuming that we're going to allow users to bring their own custom constraints). Constraints are uniquely identified by their respective ISL keywords. The fact that the `ConstraintId` introduces a self-referential type parameter into `Constraint` is intentional, and I don't think it adds any cognitive load unless you are trying to implement your own constraint. If you are trying to implement a custom constraint, then it does increase the cognitive load, but I think that can be mitigated with better documentation, and I think the trade-off is worth it in order to have stronger guarantees about type safety.
* The static variable is not strictly necessary, but it seemed like a convenient place to put it where someone could find it.  I realize, though, that I was trying to be too clever by making the companion object itself be the ID, so I'm cutting that bit out. I could also get rid of the `@JvmStatic val ID` in favor of having all of the known constraint ids as members of a pseudo-enum `object`. Do you have an opinion on that?

Reply:
Sorry for the lateness of reply... I also regret that I don't have more time to understand better, so take this with a grain of salt...

I guess what I can't tell is why it's important that the Id tie to the type of Constraint. I posit that's not actually what's important. What I think is important is that, for an identifier you know how to construct an instance of that Constraint. That makes me think of two potential tweaks:
1. Make the ConstraintRegistration the API that implementers implement. That "has-an" Id and "has-a" Constraint.
2. But it doesn't actually "have-a" Constraint! It has the ability to construct a Constraint from the (typed) arguments you expect to specify it with.
IDK, I may be wildly off, but that's what I got rn.